fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -0,0 +1,749 @@
+/// The index value is used to decide the number of '0' needed to pad to
+/// make the parts an valid Base64 encoding.
+fn find_value(ascii_string: &str, prefix: &str, suffix: &str) -> Result<String, FastCryptoError> {
+    let start = ascii_string
+        .find(prefix)
+        .ok_or_else(|| FastCryptoError::GeneralError("Invalid parts prefix".to_string()))?
+        + prefix.len();
+    let end = ascii_string[start..]
+        .find(suffix)
+        .ok_or_else(|| FastCryptoError::GeneralError("Invalid ascii suffix".to_string()))?
+        + start;
+    Ok(ascii_string[start..end].to_string())
+}
+
+/// Count the number of 0s in the bit array and check if the count satifies as the
+/// smallest, non-negative solution to equation jwt_length + 1 + K = 448 (mod 512).
+/// See more at 4.1(b) https://datatracker.ietf.org/doc/html/rfc4634#section-4.1
+fn validate_zeros_count(arr: &[bool], jwt_length: usize) -> Result<(), FastCryptoError> {
+    // Count the number of 0s in the bitarray excluding the last 8 bytes (64 bits).
+    let count = arr.iter().take(arr.len() - 64).filter(|&bit| !bit).count();
+    if (jwt_length + 1 + count) % 512 == 448 && count < 512 {
+        Ok(())
+    } else {
+        Err(FastCryptoError::GeneralError(
+            "Invalid bitarray".to_string(),
+        ))
+    }
+}
+
+/// Map a base64 string to a bit array by taking each char's index and covert it to binary form.
+fn base64_to_bitarray(input: &str) -> Vec<bool> {
+    let base64_url_character_set =
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
+    input
+        .chars()
+        .flat_map(|c| {
+            let index = base64_url_character_set.find(c).unwrap();
+            let mut bits = Vec::new();
+            for i in 0..6 {
+                bits.push((index >> (5 - i)) & 1 == 1);
+            }
+            bits
+        })
+        .collect()
+}
+
+/// Convert a bitarray to a bytearray.
+fn bits_to_bytes(bits: &[bool]) -> Vec<u8> {
+    let mut bytes: Vec<u8> = Vec::new();
+    let mut current_byte: u8 = 0;
+    let mut bits_remaining: u8 = 8;
+
+    for bit in bits.iter() {
+        if *bit {
+            current_byte |= 1 << (bits_remaining - 1);
+        }
+        bits_remaining -= 1;
+        if bits_remaining == 0 {
+            bytes.push(current_byte);
+            current_byte = 0;
+            bits_remaining = 8;
+        }
+    }
+
+    if bits_remaining < 8 {
+        bytes.push(current_byte);
+    }
+
+    bytes
+}
+
+/// Convert a big int string to a big endian bytearray.
+pub fn big_int_str_to_bytes(value: &str) -> Vec<u8> {
+    BigInt::from_str(value)
+        .expect("Invalid big int string")
+        .to_bytes_be()
+        .1
+}
+
+/// Calculate the integer value from the bytearray.
+fn calculate_value_from_bytearray(arr: &[u8]) -> usize {
+    let sized: [u8; 8] = arr.try_into().expect("Invalid byte array");
+    ((sized[7] as u16) | (sized[6] as u16) << 8).into()
```

Comment:
Is this assuming that the length will fit into the last two bytes? That seems fine, but we should at least check that the remaining 6 bytes are zeroes. Also, if you're only using this function at one place: why not take a fixed-length array of size 8 as input?